driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2005,23 +2005,46 @@ public void refreshSchemaAndSignal(final Connection connection, final DefaultRes
             executor.submit(new Runnable() {
                 @Override
                 public void run() {
+                    boolean reachedSchemaAgreement = false;
                     try {
                         // Before refreshing the schema, wait for schema agreement so
                         // that querying a table just after having created it don't fail.
-                        if (!ControlConnection.waitForSchemaAgreement(connection, Cluster.Manager.this))
+                        reachedSchemaAgreement = ControlConnection.waitForSchemaAgreement(connection, Cluster.Manager.this);
+                        if (!reachedSchemaAgreement)
                             logger.warn("No schema agreement from live replicas after {} s. The schema may not be up to date on some nodes.", configuration.getProtocolOptions().getMaxSchemaAgreementWaitSeconds());
                         ControlConnection.refreshSchema(connection, keyspace, table, Cluster.Manager.this, false);
                     } catch (Exception e) {
                         logger.error("Error during schema refresh ({}). The schema from Cluster.getMetadata() might appear stale. Asynchronously submitting job to fix.", e.getMessage());
                         submitSchemaRefresh(keyspace, table);
                     } finally {
-                        // Always sets the result
+                        // Always sets the result, but remember if we reached schema agreement
+                        rs.getExecutionInfo().setReachedSchemaAgreement(reachedSchemaAgreement);
                         future.setResult(rs);
                     }
                 }
             });
         }
 
+        public void waitForSchemaAgreementAndSignal(final Connection connection, final DefaultResultSetFuture future, final ResultSet rs) {
```

Comment:
On a DROP, there's no need to refresh the schema, we can simply remove the dropped element from our local metadata representation. This was introduced as an optimization in the context of [JAVA-543](https://datastax-oss.atlassian.net/browse/JAVA-543).
I considered merging the two methods but even though the code is similar they serve a different purpose, so I think clarity would suffer. I'll try it to see if I can come up with something clear enough.